Azure/sonic-buildimage
# [supervisor] Prevent crash if system clock rolls back within startsecs; Refactor patch

**- What I did**
- Add a patch (and related test) to supervisor in order to prevent a crash when a process has already entered the RUNNING state but then clock rolled back to be less than `startsecs`. The crash would present itself with log messages similar to the following:

```
Traceback (most recent call last):
File "/usr/bin/supervisord", line 9, in <module>
load_entry_point('supervisor==3.3.3', 'console_scripts', 'supervisord')()
File "/usr/lib/python2.7/dist-packages/supervisor/supervisord.py", line 363, in main
go(options)
File "/usr/lib/python2.7/dist-packages/supervisor/supervisord.py", line 373, in go
d.main()
File "/usr/lib/python2.7/dist-packages/supervisor/supervisord.py", line 84, in main
self.run()
File "/usr/lib/python2.7/dist-packages/supervisor/supervisord.py", line 101, in run
self.runforever()
File "/usr/lib/python2.7/dist-packages/supervisor/supervisord.py", line 253, in runforever
self.reap()
File "/usr/lib/python2.7/dist-packages/supervisor/supervisord.py", line 289, in reap
process.finish(pid, sts)
File "/usr/lib/python2.7/dist-packages/supervisor/process.py", line 527, in finish
self._assertInState(ProcessStates.STARTING)
File "/usr/lib/python2.7/dist-packages/supervisor/process.py", line 179, in _assertInState
self.config.name, current_state, allowable_states))
AssertionError: Assertion failed for start.sh: RUNNING not in STARTING
```

- Also refactored existing patch (to prevent supervisor from waiting an excessive amount of time if the system clock rolls backward) to match my simplified PR (with additional tests) which is open in the Supervisor repo (https://github.com/Supervisor/supervisor/pull/1047)

- This PR _should_ also pick cleanly into the 201803 branch.

**- How to verify it**
- Start a one-shot process using supervisor with `startsecs` set to a value > 0. Shortly after `startsecs` elapses (which causes supervisor to place the process in the RUNNING state), roll the system clock backward by a value <= `startsecs`. Then, within the amount of seconds you just rolled the clock back, cause the process to exit. Without this fix, at this point, supervisor will crash. With this fix, supervisor should recognize that the process was already in the RUNNING state and allow the process to exit without crashing.
